Machine friendly exceptions

lwitkowski opened the pull request:
Fix #4

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/not_signed)](https://cla-assistant.io/jhipster/generator-jhipster-quarkus?pullRequest=183) <br/>Thank you for your submission! We really appreciate it. Like many open source projects, we ask that you sign our [Contributor License Agreement](https://cla-assistant.io/jhipster/generator-jhipster-quarkus?pullRequest=183) before we can accept your contribution.<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla-assistant.io/check/jhipster/generator-jhipster-quarkus?pullRequest=183) it.</sub>

danielpetisme commented:
That's so awesome!
Regarding tests, you're right we don't have any tests to check the Problem response.
Feel free to add tests to the Java base code if you think it makes sense.

The simpler test you can add is ensuring the files are properly generated
https://github.com/jhipster/generator-jhipster-quarkus/blob/main/test/server.spec.js

danielpetisme commented on generators/server/templates/quarkus/src/main/java/package/web/rest/errors/EmailAlreadyUsedException.java.ejs:
```diff
@@ -18,9 +18,13 @@
+
 public class EmailAlreadyUsedException extends BadRequestAlertException {
 
+    private static final URI TYPE = URI.create(ErrorConstants.PROBLEM_BASE_URL + "/email-already-used");
```
Can you move `TYPE` to `ErrorConstants`, please

lwitkowski replied:
Sure I can, but let me first explain, why I think it's better this way: those specific exceptions are included on condition (!skipUserManagement) - if they are not included, people will see some strange **unused** constants in `ErrorConstants`. I also don't think they need to be shared, and even if - I'd prefer little duplication over coupling and making this conditional inclusion harder. 
But I may be missing some bigger picture here, so I'll happily move them back to `ErrorConstants` class :)

danielpetisme replied:
🤔 I tend to try to keep consistency between JHipster Spring and Quarkus that's why I would recommend using `ErrorConstants`. But I understand your argument...
Let it as is.

danielpetisme commented on generators/server/templates/quarkus/src/main/java/package/web/rest/errors/InvalidPasswordWebException.java.ejs:
```diff
@@ -18,19 +18,13 @@
 
-public class UserNotAuthenticatedException extends WebApplicationException {
-    private static final long serialVersionUID = 1L;
+    public static final URI TYPE = URI.create(ErrorConstants.PROBLEM_BASE_URL + "/invalid-password");
```
Same here

danielpetisme commented on generators/server/templates/quarkus/src/main/java/package/web/rest/errors/LoginAlreadyUsedException.java.ejs:
```diff
@@ -18,9 +18,13 @@
+
 public class LoginAlreadyUsedException extends BadRequestAlertException {
 
+    public static final URI TYPE = URI.create(ErrorConstants.PROBLEM_BASE_URL + "/login-already-used");
```
Same

danielpetisme commented:
Do you plan to add `ExceptionTranslator` also?
The translator adds some headers to the HTTP response on top of the Problem payload.
Again thank you for your contribution!

lwitkowski commented:
> Do you plan to add `ExceptionTranslator` also?

Actually I don't :) All headers are added in `BadRequestAlertException` class and as far as I can see it just works. I find this solution cleaner than `ExceptionTranslator`.

The thing is Zalando's Problem implementation does not have notion of http header - there's no such thing as ProblemWithHeaders, so by default you may have either `Problem`, or customer headers (by using `WebApplicationException`). 

`BadRequestAlertException` is kind of a walk around to have both `application/problem+json` body and custom headers, but it's not generic solution for all exceptions.

Does that make sense?

danielpetisme commented:
Ok make sense now, smart analysis.

lwitkowski commented:
Thanks for first review @danielpetisme, I'll then focus on adding some tests, and will then mark this PR as 'ready for review'.

lwitkowski commented:
OK, I added some tests, but not everything is super clear for me which exceptions should have additional headers and which not necessarily, but I guess it's not critical - this version is closer to 'reference' SpringBoot version for sure :)

One more note: `quarkus-resteasy-problem` extension logs all exceptions/problems that flows through it's exception mappers by default, and it looks like this:
`2021-04-08 21:18:07,825 INFO  [http-problem] (executor-thread-1) status=404, title="Not Found", detail="HTTP 404 Not Found"`
We can easily silence them by setting `quarkus.log.category.http-problem.level` to ERROR or OFF - not sure which default is better for this project.

lwitkowski commented on generators/server/templates/quarkus/src/test/java/package/web/rest/UserResourceTest.java.ejs:
```diff
@@ -162,7 +166,15 @@ public class UserResourceTest {
+            .header("X-" + applicationName + "-params", is("userManagement"))
+            .contentType("application/problem+json")
+            .body("title", is("A new user cannot already have an ID"))
+            .body("entityName", is("userManagement"))
```
I'm also not super sure if all those fields (entityName, errorKey, message, params) are needed/used by frontend - I used springboot version as reference and added everything I could find there.

danielpetisme replied:
They do

danielpetisme commented:
> OK, I added some tests, but not everything is super clear for me which exceptions should have additional headers and which not necessarily, but I guess it's not critical - this version is closer to 'reference' SpringBoot version for sure :)
> 
> One more note: `quarkus-resteasy-problem` extension logs all exceptions/problems that flows through it's exception mappers by default, and it looks like this:
> `2021-04-08 21:18:07,825 INFO [http-problem] (executor-thread-1) status=404, title="Not Found", detail="HTTP 404 Not Found"`
> We can easily silence them by setting `quarkus.log.category.http-problem.level` to ERROR or OFF - not sure which default is better for this project.

I propose to let it as is for now.

lwitkowski commented on generators/server/templates/quarkus/src/main/java/package/web/util/ResponseUtil.java.ejs:
```diff
@@ -18,21 +18,25 @@
 public interface ResponseUtil {
-    static Response wrapOrNotFound(Optional maybeResponse) {
-        return wrapOrNotFound(maybeResponse, Collections.emptyMap());
+    static Response wrapOrNotFound(Optional<?> maybeResponse) {
```
You're right, I'll try to minimise the changes to limit the drift from reference spring impl.

lwitkowski replied:
Fixed.

lwitkowski commented on generators/server/templates/quarkus/src/test/java/package/web/rest/AccountResourceTest_jwt.java.ejs:
```diff
@@ -292,7 +300,20 @@ public class AccountResourceTest {
+            .body("violations", hasSize(2))
+            .body("violations.field", hasItems("email", "password"))
+            .body("violations.error", hasItems("must be a well-formed email address", "size must be between 4 and 100"));
+            // Springboot version returns a bit different schema. Does frontend care/use/display fieldErrors?
```
Spring version is not perfect (camelCase vs snakcase), `violations` is more in line with zalando problem violations library impl, but I'll add adjust to it, and maybe propose change to reference project first :)

lwitkowski commented:
@danielpetisme I've addressed all comments (hopefully :), I also updated quarkus-resteasy-problem extension to the latest stable version (which e.g supports quarkus-reactive now - could be relevant for future of this project).

danielpetisme commented:
LGTM, @DanielFran do you see anything else to update?

DanielFran commented:
@danielpetisme No, seems OK for me.

danielpetisme commented:
Thank you so much @lwitkowski !
Please, claim the associated bounty
https://www.jhipster.tech/bug-bounties/

DanielFran commented:
Congratulation @lwitkowski for the contribution, well done.

lwitkowski commented:
That's awesome, thanks for help @danielpetisme and @DanielFran. I hope it's not my last contribution :)

danielpetisme commented:
We hope so too 🙂

avdev4j commented:
thank you @lwitkowski !

I will be happy to see you again :D

lwitkowski commented:
Here's bounty claim: https://opencollective.com/generator-jhipster/expenses/38542
